File: src/bmsx/render/vdp/source_pixels.ts
import type { VdpHostOutput } from '../../machine/devices/vdp/vdp';
import { resolveVdpHostSurfaceSlot } from './surfaces';

export type VdpSurfacePixels = {
	pixels: Uint8Array;
	width: number;
	height: number;
	stride: number;
};

export function resolveVdpSurfacePixels(output: VdpHostOutput, surfaceId: number): VdpSurfacePixels {
	const slot = resolveVdpHostSurfaceSlot(output, surfaceId);
	return {
		pixels: slot.cpuReadback,
		width: slot.surfaceWidth,
		height: slot.surfaceHeight,
		stride: slot.surfaceWidth * 4,
	};
}

File: src/bmsx_cpp/render/vdp/source_pixels.cpp
#include "render/vdp/source_pixels.h"

#include "render/vdp/surfaces.h"

namespace bmsx {

VdpSurfacePixels resolveVdpSurfacePixels(const VDP::VdpHostOutput& output, u32 surfaceId) {
	const VDP::VramSlot& slot = resolveVdpHostSurfaceSlot(output, surfaceId);
	return VdpSurfacePixels{
		slot.cpuReadback.data(),
		slot.surfaceWidth,
		slot.surfaceHeight,
		slot.surfaceWidth * 4u,
	};
}

} // namespace bmsx
